### PR TITLE
IP endpoint for dnstap

### LINKED ISSUE #924: mw/dnstap: IP endpoint for dnstap
In a microservices environment, it makes more sense to send dnstap data over the network rather than to a domain socket. What level of effort would that be?

### PR DESCRIPTION
**Pull request by 2buisson**
Fixes #924 thanks to https://github.com/dnstap/golang-dnstap/pull/13.

### CODE REVIEW

**Comment by yongtang** [comment]
I think the common way to define unix vs. tcp socket is to specify protocol:
```
unix:///var/run/dnstap.sock
tcp://127.0.0.1:8080
```

Maybe we could add the protocol at the beginning (and default to `unix://` if it is not specified)?

File: middleware/dnstap/README.md [hunk 1]
```diff
@@ -27,6 +27,12 @@ Log information including the wire-format DNS message about client requests and
 dnstap /tmp/dnstap.sock full
 ~~~
 
+Log to a remote endpoint.
+
+~~~ txt
+dnstap 127.0.0.1:6000 full
```
Comment (yongtang) [inline middleware/dnstap/README.md]:
```
unix:///var/run/dnstap.sock
tcp://127.0.0.1:8080
```

**Review by yongtang — CHANGES_REQUESTED** [review]

File: middleware/dnstap/out/tcp.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+type TCP struct {
+	address string
+	frames  [][]byte
+}
+
+// NewTCP returns a TCP writer.
+func NewTCP(address string) (s *TCP) {
+	s = &TCP{address: address}
+	s.frames = make([][]byte, 0, 13) // 13 messages buffer
+	return
```
Comment (miekg) [inline middleware/dnstap/out/tcp.go]: general observation: don't do these naked returns :-)

File: middleware/dnstap/out/tcp.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	for _, frame := range s.frames {
+		if _, err = enc.Write(frame); err != nil {
+			return
+		}
+	}
+	err = enc.Flush()
+	return
+}
+
+// Close is an alias to Flush.
```
Comment (miekg) [inline middleware/dnstap/out/tcp.go]: Why is this needed?

File: middleware/dnstap/out/tcp_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	if err := tcp.Flush(); err != nil {
+		return err
+	}
+	return nil
+}
+func TestTcp(t *testing.T) {
+	tcp := NewTCP("localhost:14000")
+
+	if err := sendOneTcp(tcp); err == nil {
+		t.Fatal("not listening but no error")
```
Comment (miekg) [inline middleware/dnstap/out/tcp_test.go]: Test messages are different from errors, as they will not be wrapped and will be directly read by humans. So make them a full sentence "Not listening, but no error."
Why is this an Fatal in the test btw?

File: middleware/dnstap/setup.go [hunk 5]
```diff
@@ -26,30 +28,43 @@ func wrapSetup(c *caddy.Controller) error {
 	return nil
 }
 
-func parseConfig(c *caddyfile.Dispenser) (path string, full bool, err error) {
+func parseConfig(c *caddyfile.Dispenser) (path string, socket, full bool, err error) {
```
Comment (miekg) [inline middleware/dnstap/setup.go]: Shouldn't keep adding return parameters in this function. Either start an dnsOpt struct which has all these or, if possible, fill out a *dnstap structure with the correct elements.

File: middleware/dnstap/setup.go [hunk 6]
```diff
@@ -26,30 +28,43 @@ func wrapSetup(c *caddy.Controller) error {
...
 	dnstap := Dnstap{Pack: full}
 
-	o, err := out.NewSocket(path)
-	if err != nil {
-		log.Printf("[WARN] Can't connect to %s at the moment", path)
+	var o io.WriteCloser
+	if socket {
+		o, err = out.NewSocket(path)
+		if err != nil {
+			log.Printf("[WARN] Can't connect to %s at the moment", path)
```
Comment (miekg) [inline middleware/dnstap/setup.go]: Include error in log message.

File: middleware/dnstap/out/tcp.go (hunk 3 above)
Comment (2buisson) [reply middleware/dnstap/out/tcp.go]: To satisfy `io.WriteCloser` similarly to type `out.Socket`.

**Review by yongtang — APPROVED** [review]

File: middleware/dnstap/out/tcp.go (hunk 3 above)
Comment (miekg) [reply middleware/dnstap/out/tcp.go]: Ok, please add that to the comment :)
